Documentation/Guides/SSL Pinning.md

```diff
@@ -0,0 +1,36 @@
+
+`04 B9 ED ED FC C4 3D 5D 2B C6 64 C1 53 A3 B6 52 56 7B BF E2 86 AB FF 94 97 FD 9C CA 11 3B 23 2D 8B 74 7B 94 7F 99 00 13 FA D1 01 73 00 B3 90 3B 74 F2 20 D7 47 53 D4 F8 33 58 A5 07 F5 12 39 8A 7E`
+
+These bytes can then be run through a SHA-256 hash (an easy way to do this is using the `CertificateHasher` in Hyperspace) to produce your pinning hash: `XlCPXC6IrttTF9Y1887kS+efCCf3uFjHW6D1TUI9f+Q=`. This hash can then be passed into a `DomainConfiguration` using `.init(domain:pinningHashes:)` or `.init(domain:encodedPinningHashes:)` (the latter of which is Base64 encoded). Please note that the hash required for the pinning implementation in Hyperspace is not identical to the one that the `NSPinnedDomains` key uses in iOS 14+.
+
+Once the `TrustConfiguration` has been created with a `[DomainConfiguration]`, the final step is to tell Hyperspace how to validate this configuration at request-time. This can be done by manually utilizing a `TrustValidator` at the appropriate time, and then by calling `validator.handle(challenge:,handler:)`. In the vast majority of cases though, it is simpler to use the provided `TrustValidatingTransportService` which builds upon `TransportService` to automatically provide this functionality when the app receives an authentication challenge through the `URLSessionDelegate` for an appropriate domain.
+
+Once this is in place you should be able to observe the effects of pinning, for example using a proxy tool like Charles. When the proxy is not running, the certificate should be the "real" one and any requests you make to the pinned domain should succeed. Turning on the proxy will instead use a different proxy certificate, which will cause any requests to the pinned domain to return an error, similar to the one below:
+
+
+``` 
+failure(TransportFailure(error: Hyperspace.TransportError(code: Hyperspace.TransportError.Code.cancelled, underlyingError: Optional(Error Domain=NSURLErrorDomain Code=-999 \"cancelled\" UserInfo={NSErrorFailingURLStringKey=https://jsonplaceholder.typicode.com/users/1, NSErrorFailingURLKey=https://jsonplaceholder.typicode.com/users/1, _NSURLErrorRelatedURLSessionTaskErrorKey=(\n    \"LocalDataTask <AA005DD8-5591-4437-86E7-210A92624C73>.<1>\"\n), _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <AA005DD8-5591-4437-86E7-210A92624C73>.<1>, NSLocalizedDescription=cancelled})), request: Hyperspace.HTTP.Request(url: Optional(https://jsonplaceholder.typicode.com/users/1), method: Optional(\"GET\"), headers: Optional([:]), body: nil), response: nil))"
+```
```

Comment:
It could be good to have another section in here talking about the potential tradeoffs/considerations when implementing SSL pinning. For example, what happens when there is a failure or something happens and the certificates change without proper planning. The main thing to highlight would probably be that pinning failures mean that no network requests go through, which would likely provide a "broken" end user experience.

Reply:
Agreed, I added some additional considerations - let me know if that is similar to what you were thinking.